srml/support/procedural/src/storage/transfo.rs

```diff
@@ -0,0 +1,619 @@
+			let mutate_impl = if !is_option {
+				quote!{
+					<Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::insert(key, &val, storage)
+				}
+			} else {
+				quote!{
+					match val {
+						Some(ref val) => <Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::insert(key, &val, storage),
+						None => <Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::remove(key, storage),
+					}
+				}
+			};
+			// generator for map
+			quote!{
+				#visibility struct #name<#traitinstance: #traittype>(#scrate::storage::generator::PhantomData<#traitinstance>);
+
+				impl<#traitinstance: #traittype> #scrate::storage::generator::StorageMap<#kty, #typ> for #name<#traitinstance> {
+					type Query = #gettype;
+
+					/// Get the prefix key in storage.
+					fn prefix() -> &'static [u8] {
+						stringify!(#cratename #name).as_bytes()
+					}
+
+					/// Get the storage key used to fetch a value corresponding to a specific key.
+					fn key_for(x: &#kty) -> #scrate::rstd::vec::Vec<u8> {
+						let mut key = <#name<#traitinstance> as #scrate::storage::generator::StorageMap<#kty, #typ>>::prefix().to_vec();
+						#scrate::codec::Encode::encode_to(x, &mut key);
+						key
+					}
+
+					/// Load the value associated with the given key from the map.
+					fn get<S: #scrate::GenericStorage>(key: &#kty, storage: &S) -> Self::Query {
+						let key = <#name<#traitinstance> as #scrate::storage::generator::StorageMap<#kty, #typ>>::key_for(key);
+						storage.get(&key[..]).#option_simple_1(|| #fielddefault)
+					}
+
+					/// Take the value, reading and removing it.
+					fn take<S: #scrate::GenericStorage>(key: &#kty, storage: &S) -> Self::Query {
+						let key = <#name<#traitinstance> as #scrate::storage::generator::StorageMap<#kty, #typ>>::key_for(key);
+						storage.take(&key[..]).#option_simple_1(|| #fielddefault)
+					}
+
+					/// Mutate the value under a key
+					fn mutate<R, F: FnOnce(&mut Self::Query) -> R, S: #scrate::GenericStorage>(key: &#kty, f: F, storage: &S) -> R {
+						let mut val = <Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::take(key, storage);
+
+						let ret = f(&mut val);
+						#mutate_impl ;
+						ret
+					}
+
+				}
+
+			}
+		};
+		impls.push(implementation)
+	}
+
+	quote!( #( #impls )* )
+
+}
+
+
+fn decl_store_items(
+	storage_lines: &ext::Punctuated<DeclStorageLine, Token![;]>,
+) -> proc_macro2::TokenStream {
+	let mut items = Vec::new();
+	for sline in storage_lines.inner.iter() {
+		let DeclStorageLine {
+			name,
+			..
+		} = sline;
+		items.push(quote!{
+			type #name;
+		});
+	}
+
+	quote!( #( #items )* )
+}
+
+fn impl_store_items(
+	traitinstance: &Ident,
+	storage_lines: &ext::Punctuated<DeclStorageLine, Token![;]>,
+) -> proc_macro2::TokenStream {
+	let mut items = Vec::new();
+	for sline in storage_lines.inner.iter() {
+		let DeclStorageLine {
+			name,
+			..
+		} = sline;
+		items.push(quote!{
+			type #name = #name<#traitinstance>;
```

Comment:
Same as above and declare `let traitinstance = std::iter::repeat(traitinstance)`